geode-core/src/main/java/org/apache/geode/internal/cache/TXState.java

```diff
@@ -95,6 +101,15 @@
    */
   private int modSerialNum;
   private final List<EntryEventImpl> pendingCallbacks = new ArrayList<EntryEventImpl>();
+  /**
+   * for client/server JTA transactions we need to have a single thread handle both beforeCompletion
+   * and afterCompletion so that beforeCompletion can obtain locks for the afterCompletion step.
+   * This is that thread
+   */
+  protected volatile TXStateSynchronizationRunnable syncRunnable;
+
+  private volatile SynchronizationCommitConflictException beforeCompletionException;
```

Comment:
I think these should be moved to TXStateSynchronizationRunnable and probably do not need to be volatile on it. We can talk about this.